src/keystore.h

```diff
@@ -19,21 +15,28 @@ class CScriptID;
 /** A virtual base class for key stores */
 class CKeyStore
 {
-protected:
-    mutable CCriticalSection cs_KeyStore;
-
 public:
     virtual ~CKeyStore() {}
 
     //! Add a key to the store.
     virtual bool AddKeyPubKey(const CKey &key, const CPubKey &pubkey) =0;
-    virtual bool AddKey(const CKey &key);
+    bool AddKey(const CKey &key) 
+    {
+        return AddKeyPubKey(key, key.GetPubKey());
+    }
 
     //! Check whether a key corresponding to a given address is present in the store.
     virtual bool HaveKey(const CKeyID &address) const =0;
     virtual bool GetKey(const CKeyID &address, CKey& keyOut) const =0;
     virtual void GetKeys(std::set<CKeyID> &setAddress) const =0;
-    virtual bool GetPubKey(const CKeyID &address, CPubKey& vchPubKeyOut) const;
+    bool GetPubKey(const CKeyID &address, CPubKey& vchPubKeyOut) const
+    {
+        CKey key;
+        if (!GetKey(address, key))
+            return false;
+        vchPubKeyOut = key.GetPubKey();
+        return true;
+    }
 
     //! Support for BIP 0013 : see https://github.com/bitcoin/bips/blob/master/bip-0013.mediawiki
     virtual bool AddCScript(const CScript& redeemScript) =0;
```

Comment:
Sorry for not answering earlier. If you comment those methods you'll get this errors:
```
wallet_ismine.cpp: In function ‘isminetype IsMine(const CKeyStore&, const CScript&)’:
wallet_ismine.cpp:42:22: error: ‘const class CKeyStore’ has no member named ‘HaveWatchOnly’
         if (keystore.HaveWatchOnly(scriptPubKey))
                      ^
wallet_ismine.cpp:88:18: error: ‘const class CKeyStore’ has no member named ‘HaveWatchOnly’
     if (keystore.HaveWatchOnly(scriptPubKey))
```
The other 3 methods can be safely removed from the interface.
I mean, it's not unsolvable but I wanted to keep this simple...